code/game/objects/effects/effect_system/fluid_spread/effects_smoke.dm

```diff
@@ -0,0 +1,421 @@
+/**
+ * A fluid which spreads through the air affecting every mob it engulfs.
+ */
+/obj/effect/particle_effect/fluid/smoke
+	name = "smoke"
+	icon = 'icons/effects/96x96.dmi'
+	icon_state = "smoke"
+	pixel_x = -32
+	pixel_y = -32
+	opacity = TRUE
+	plane = ABOVE_GAME_PLANE
+	layer = FLY_LAYER
+	anchored = TRUE
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	animate_movement = FALSE
+	/// How long the smoke sticks around before it dissipates.
+	var/lifetime = 10 SECONDS
+
+/obj/effect/particle_effect/fluid/smoke/Initialize(mapload, datum/fluid_group/group, ...)
+	. = ..()
+	create_reagents(1000)
+	START_PROCESSING(SSobj, src)
+	setDir(pick(GLOB.cardinals))
+
+/obj/effect/particle_effect/fluid/smoke/Destroy()
+	STOP_PROCESSING(SSobj, src)
+	return ..()
+
+
+/**
+ * Makes the smoke fade out and then deletes it.
+ */
+/obj/effect/particle_effect/fluid/smoke/proc/kill_smoke()
+	STOP_PROCESSING(SSobj, src)
+	INVOKE_ASYNC(src, .proc/fade_out)
+	QDEL_IN(src, 1 SECONDS)
+
+/**
+ * Animates the smoke gradually fading out of visibility.
+ * Also makes the smoke turf transparent as it passes 160 alpha.
+ *
+ * Arguments:
+ * - frames = 0.8 [SECONDS]: The amount of time the smoke should fade out over.
+ */
+/obj/effect/particle_effect/fluid/smoke/proc/fade_out(frames = 0.8 SECONDS)
+	if(alpha == 0) //Handle already transparent case
+		if(opacity)
+			set_opacity(FALSE)
+		return
+
+	if(frames == 0)
+		set_opacity(FALSE)
+		alpha = 0
+		return
+
+	var/time_to_transparency = round(((alpha - 160) / alpha) * frames)
+	if(time_to_transparency >= 1)
+		addtimer(CALLBACK(src, /atom.proc/set_opacity, FALSE), time_to_transparency)
+	else
+		set_opacity(FALSE)
+	animate(src, time = frames, alpha = 0)
+
+
+/obj/effect/particle_effect/fluid/smoke/spread()
+	if(group.total_size > group.target_size)
+		return
+	var/turf/t_loc = get_turf(src)
+	if(!t_loc)
+		return
+
+	for(var/turf/spread_turf in t_loc.get_atmos_adjacent_turfs())
+		if(group.total_size > group.target_size)
+			break
+		if(locate(type) in spread_turf)
+			continue // Don't spread smoke where there's already smoke!
+		for(var/mob/living/smoker in spread_turf)
+			smoke_mob(smoker, SSOBJ_DT)
+
+		var/obj/effect/particle_effect/fluid/smoke/spread_smoke = new type(spread_turf, group)
+		reagents.copy_to(spread_smoke, reagents.total_volume)
+		spread_smoke.add_atom_colour(color, FIXED_COLOUR_PRIORITY)
+		spread_smoke.lifetime = lifetime
+
+		// the smoke spreads rapidly, but not instantly
+		addtimer(CALLBACK(spread_smoke, /obj/effect/particle_effect/fluid.proc/spread), 1)
+
+
+/obj/effect/particle_effect/fluid/smoke/process(delta_time)
+	lifetime -= delta_time SECONDS
+	if(lifetime <= 0)
+		kill_smoke()
+		return FALSE
+	for(var/mob/living/smoker in range(0,src))
```

Comment:
isn't this in get_turf(src)

Reply:
I think so. It also gets the smoke itself and everything in the smoke... but if you have atoms in your smoke something has gone weird.

Reply:
Changed it to `loc` in one of the commits. That way it should be basically `get_turf(src)` in most situations, but if someone ever manages to get smoke into a locker it should still work.